### PR TITLE
Parse Postgres interval strings with Zod

### PR DESCRIPTION
**Pull request by nwalters512**
I originally tried to do this with the new Temporal API, but Postgres has some quirks that make this infeasible. Namely, Postgres tracks the sign separately for years, months, days, and the time, which doesn't leave us with a straightforward way to convert them into `Temporal.Duration` objects.

Ultimately, I opted to match the behavior one gets when doing `EXTRACT(epoch from '...'::interval)` for an interval, which is what we've been doing in PrairieTest when converting from PG to JS data types (see the `interval_to_milliseconds` sproc).

### CODE REVIEW

File: apps/prairielearn/src/lib/db-types.ts [hunk 1]
```diff
@@ -7,7 +15,25 @@ import { z } from 'zod';
...
-  .refine((val) => /^\d+$/.test(val), { message: 'ID is not an integer' });
+  .refine((val) => /^\d+$/.test(val), { message: 'ID is not a non-negative integer' });
+
+export const IntervalSchema = z.string().transform((val) => {
+  const interval = parsePostgresInterval(val);
+
+  // This calculation matches Postgres's behavior when computing the number of
+  // milliseconds in an interval with `EXTRACT(epoch from '...'::interval) * 1000`.
+  // The noteworthy parts of this conversion are that 1 year = 365.25 days and
+  // 1 month = 30 days.
```
Comment (nwalters512) [inline apps/prairielearn/src/lib/db-types.ts:26]: I wasn't actually able to find this documented explicitly in the Postgres docs, but I found references to this fact [on StackOverflow](https://stackoverflow.com/questions/952493/how-do-i-convert-an-interval-into-a-number-of-hours-with-postgres) and confirmed this experimentally.

File: apps/prairielearn/src/lib/db-types.test.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  it('rejects a non-numeric ID', () => {
+    const result = IdSchema.safeParse('abc');
+    assert.isFalse(result.success);
+  });
+});
+
+describe('IntervalSchema', () => {
+  it('parses an interval with date', () => {
+    const interval = IntervalSchema.parse('1 year 2 months 3 days');
+    assert.equal(interval, 37000800000);
```
Comment (nwalters512) [inline apps/prairielearn/src/lib/db-types.test.ts:35]: I generated this and all the other expected values by running the intervals through `SELECT EXTRACT(epoch FROM ...)`.

**Review by mwest1066 — APPROVED** [review]
Awesome! We need this in PT too. Do you want to just copy/paste it, or stick this (and `IdSchema`, etc) in a package?

**Comment by nwalters512** [comment]
Let's copy/paste for now, if that's good with you. I suspect we might need to adjust this slightly once we start using it in PT, and it'd be good to be able to quickly make local changes there as needed.

**Pull request merged**